esModuleInterop conflates default import type with that of a member named "default"

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨 -->
<!--
Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker.

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the CONTRIBUTING guidelines: https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ
-->

<!-- If you have a QUESTION:
   THIS IS NOT A FORUM FOR QUESTIONS.
   Ask questions at http://stackoverflow.com/questions/tagged/typescript
    or https://gitter.im/Microsoft/TypeScript
-->

<!-- If you have a SUGGESTION:
  Most suggestion reports are duplicates, please search extra hard before logging a new suggestion.
  See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md
-->

<!-- If you have a BUG:
  Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  2.8.0-dev.20180322

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** 
`esModuleInterop`

**Code**

The flag `--esModuleInterop` must be enabled.

`./index.ts`

```ts
import Point from "./point";

const C = Point;
const p = new C(1, 2);
```

`./point/index.js`

```js
function Point(x, y) {
  this.x = x;
  this.y = y;
}

Point.default = "foo";

module.exports = Point;
```

`./point/index.d.ts`

```ts
declare class Point {
  x: number;
  y: number;

  constructor(x: number, y: number);

  static default: "foo";
}

export = Point;
```
**Expected behavior:**

The code compiles.

**Actual behavior:**

```
index.ts(4,11): error TS2351: Cannot use 'new' with an expression whose type lacks a call or construct signature.
```

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> 

This is not reproducible on Playground.

**Additional Information:**

The static member `default` causes TypeScript to mistake the type of the imported member `Point` as that of the static member, which is `"foo"` in this case, rather than the class Point, which is approximately the merger of `interface Point { x: number; y: number}` and `var Point: { new(x: number, y: number): Point; default: "foo"; }`.

This impacts using `yargs` with `@types/yargs` in the wild.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -1730,14 +1730,25 @@ namespace ts {
                 : resolveSymbol(moduleSymbol.exports.get(name), dontResolveAlias);
         }
 
+        function isSyntacticDefault(node: Node) {
+            if (isExportAssignment(node) && !node.isExportEquals) {
+                return true;
+            }
+            if (hasModifier(node, ModifierFlags.Default)) {
+                return true;
+            }
+            return false;
+        }
+
         function canHaveSyntheticDefault(file: SourceFile | undefined, moduleSymbol: Symbol, dontResolveAlias: boolean) {
             if (!allowSyntheticDefaultImports) {
                 return false;
             }
             // Declaration files (and ambient modules)
             if (!file || file.isDeclarationFile) {
-                // Definitely cannot have a synthetic default if they have a default member specified
-                if (resolveExportByName(moduleSymbol, InternalSymbolName.Default, dontResolveAlias)) {
+                // Definitely cannot have a synthetic default if they have a syntactic default member specified
+                const defaultExportSymbol = resolveExportByName(moduleSymbol, InternalSymbolName.Default, dontResolveAlias);
+                if (defaultExportSymbol && defaultExportSymbol.valueDeclaration && isSyntacticDefault(defaultExportSymbol.valueDeclaration)) {
                     return false;
                 }
                 // It _might_ still be incorrect to assume there is no __esModule marker on the import at runtime, even if there is no `default` member
@@ -1777,7 +1788,7 @@ namespace ts {
                 if (!exportDefaultSymbol && !hasSyntheticDefault) {
                     error(node.name, Diagnostics.Module_0_has_no_default_export, symbolToString(moduleSymbol));
                 }
-                else if (!exportDefaultSymbol && hasSyntheticDefault) {
+                else if (hasSyntheticDefault) {
                     // per emit behavior, a synthetic default overrides a "real" .default member if `__esModule` is not present
                     return resolveExternalModuleSymbol(moduleSymbol, dontResolveAlias) || resolveSymbol(moduleSymbol, dontResolveAlias);
                 }

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -1731,13 +1731,7 @@ namespace ts {
         }
 
         function isSyntacticDefault(node: Node) {
-            if (isExportAssignment(node) && !node.isExportEquals) {
-                return true;
-            }
-            if (hasModifier(node, ModifierFlags.Default)) {
-                return true;
-            }
-            return false;
+            return ((isExportAssignment(node) && !node.isExportEquals) || hasModifier(node, ModifierFlags.Default));
         }
 
         function canHaveSyntheticDefault(file: SourceFile | undefined, moduleSymbol: Symbol, dontResolveAlias: boolean) {

--- a/tests/baselines/reference/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.js
+++ b/tests/baselines/reference/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.js
@@ -1,0 +1,29 @@
+//// [tests/cases/compiler/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.ts] ////
+
+//// [point.d.ts]
+declare class Point {
+    x: number;
+    y: number;
+
+    constructor(x: number, y: number);
+
+    static default: "foo";
+}
+
+export = Point;
+//// [index.ts]
+import Point from "./point";
+
+const C = Point;
+const p = new C(1, 2);
+
+
+//// [index.js]
+"use strict";
+var __importDefault = (this && this.__importDefault) || function (mod) {
+    return (mod && mod.__esModule) ? mod : { "default": mod };
+};
+exports.__esModule = true;
+var point_1 = __importDefault(require("./point"));
+var C = point_1["default"];
+var p = new C(1, 2);

--- a/tests/baselines/reference/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.symbols
+++ b/tests/baselines/reference/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.symbols
@@ -1,0 +1,33 @@
+=== tests/cases/compiler/point.d.ts ===
+declare class Point {
+>Point : Symbol(Point, Decl(point.d.ts, 0, 0))
+
+    x: number;
+>x : Symbol(Point.x, Decl(point.d.ts, 0, 21))
+
+    y: number;
+>y : Symbol(Point.y, Decl(point.d.ts, 1, 14))
+
+    constructor(x: number, y: number);
+>x : Symbol(x, Decl(point.d.ts, 4, 16))
+>y : Symbol(y, Decl(point.d.ts, 4, 26))
+
+    static default: "foo";
+>default : Symbol(Point.default, Decl(point.d.ts, 4, 38))
+}
+
+export = Point;
+>Point : Symbol(Point, Decl(point.d.ts, 0, 0))
+
+=== tests/cases/compiler/index.ts ===
+import Point from "./point";
+>Point : Symbol(Point, Decl(index.ts, 0, 6))
+
+const C = Point;
+>C : Symbol(C, Decl(index.ts, 2, 5))
+>Point : Symbol(Point, Decl(index.ts, 0, 6))
+
+const p = new C(1, 2);
+>p : Symbol(p, Decl(index.ts, 3, 5))
+>C : Symbol(C, Decl(index.ts, 2, 5))
+

--- a/tests/baselines/reference/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.types
+++ b/tests/baselines/reference/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.types
@@ -1,0 +1,36 @@
+=== tests/cases/compiler/point.d.ts ===
+declare class Point {
+>Point : Point
+
+    x: number;
+>x : number
+
+    y: number;
+>y : number
+
+    constructor(x: number, y: number);
+>x : number
+>y : number
+
+    static default: "foo";
+>default : "foo"
+}
+
+export = Point;
+>Point : Point
+
+=== tests/cases/compiler/index.ts ===
+import Point from "./point";
+>Point : typeof Point
+
+const C = Point;
+>C : typeof Point
+>Point : typeof Point
+
+const p = new C(1, 2);
+>p : Point
+>new C(1, 2) : Point
+>C : typeof Point
+>1 : 1
+>2 : 2
+

--- a/tests/cases/compiler/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.ts
+++ b/tests/cases/compiler/esModuleInteropDefaultMemberMustBeSyntacticallyDefaultExport.ts
@@ -1,0 +1,17 @@
+// @esModuleInterop: true
+// @filename: point.d.ts
+declare class Point {
+    x: number;
+    y: number;
+
+    constructor(x: number, y: number);
+
+    static default: "foo";
+}
+
+export = Point;
+// @filename: index.ts
+import Point from "./point";
+
+const C = Point;
+const p = new C(1, 2);

